File: Xit/Document/XTDocument.m
#import "XTDocument.h"
#import "Xit-Swift.h"
#include "XTQueueUtils.h"


@interface XTDocument ()

@property (readwrite) XTRepository *repository;

@end



@implementation XTDocument

- (instancetype)initWithType:(NSString *)typeName
                       error:(NSError * _Nullable __autoreleasing *)outError
{
  return nil;
}

- (instancetype)initWithContentsOfURL:(NSURL *)absoluteURL
                     ofType:(NSString *)typeName
                      error:(NSError **)outError
{
  self =
      [super initWithContentsOfURL:absoluteURL ofType:typeName error:outError];
  if (self) {
    _repoURL = absoluteURL;
    _repository = [[XTRepository alloc] initWithURL:_repoURL];
  }
  [(AppDelegate*)NSApp.delegate dismissOpenPanel];
  return self;
}

- (instancetype)initForURL:(NSURL *)absoluteDocumentURL
    withContentsOfURL:(NSURL *)absoluteDocumentContentsURL
               ofType:(NSString *)typeName
                error:(NSError **)outError
{
  return [self initWithContentsOfURL:absoluteDocumentURL
                              ofType:typeName
                               error:outError];
}

- (void)makeWindowControllers
{
  NSStoryboard *storyboard = [NSStoryboard storyboardWithName:@"XTDocument"
                                                       bundle:nil];
  
  XTWindowController *controller = (XTWindowController*)
      [storyboard instantiateInitialController];

  [self addWindowController:controller];
}

- (BOOL)readFromURL:(NSURL *)absoluteURL
             ofType:(NSString *)typeName
              error:(NSError **)outError
{
  NSURL *gitURL = [absoluteURL URLByAppendingPathComponent:@".git"];

  if ([[NSFileManager defaultManager] fileExistsAtPath:gitURL.path])
    return YES;

  if (outError != NULL) {
    NSDictionary *userInfo =
        @{NSLocalizedFailureReasonErrorKey :
          @"The folder does not contain a Git repository."};
    *outError = [NSError errorWithDomain:NSCocoaErrorDomain
                                    code:NSFileReadUnknownError
                                userInfo:userInfo];
  }
  return NO;
}

- (void)canCloseDocumentWithDelegate:(id)delegate
                 shouldCloseSelector:(SEL)shouldCloseSelector
                         contextInfo:(void *)contextInfo
{
  XTWindowController *controller = self.windowControllers.firstObject;
  
  [controller shutDown];
  
  [super canCloseDocumentWithDelegate:delegate
                  shouldCloseSelector:shouldCloseSelector
                          contextInfo:contextInfo];
}

- (void)updateChangeCount:(NSDocumentChangeType)change
{
  // Do nothing. There is no need for an "unsaved" state.
}

@end

File: Xit/Document/XTDocument.swift
import Foundation

class XTDocument: NSDocument
{
  private(set) var repoURL: URL!
  private(set) var repository: XTRepository!
  
  override func makeWindowControllers()
  {
    let storyboard = NSStoryboard(name: "XTDocument", bundle: nil)
    let controller = storyboard.instantiateInitialController()
                     as! XTWindowController
    
    addWindowController(controller)
  }
  
  override func read(from url: URL, ofType typeName: String) throws
  {
    let gitURL = url.appendingPathComponent(".git")
    
    guard FileManager.default.fileExists(atPath: gitURL.path)
    else {
      throw NSError(domain: NSCocoaErrorDomain, code: NSFileReadUnknownError,
                    userInfo: [NSLocalizedFailureReasonErrorKey:
                               "The folder does not contain a Git repository."])
    }
    guard let repository = XTRepository(url: url)
    else {
      throw NSError(domain: "xit", code: -1,
                    userInfo: [NSLocalizedDescriptionKey:
                               "The repository could not be opened"])
    }

    self.repoURL = url
    self.repository = repository
    
    (NSApp.delegate as! AppDelegate).dismissOpenPanel()
  }
  
  override func canClose(withDelegate delegate: Any,
                         shouldClose shouldCloseSelector: Selector?,
                         contextInfo: UnsafeMutableRawPointer?)
  {
    let controller = windowControllers.first as! XTWindowController
    
    controller.shutDown()
    
    super.canClose(withDelegate: delegate,
                   shouldClose: shouldCloseSelector,
                   contextInfo: contextInfo)
  }
  
  override func updateChangeCount(_ change: NSDocument.ChangeType)
  {
    // Do nothing. There is no need for an "unsaved" state.
  }
}
